### PR TITLE
Pattern block: Ensure consistent editing of overrides in Write Mode

### PR DESCRIPTION
**Pull request by talldan**
Based on #65204 and merges into that PR.

## What?
The pattern (core/block) block uses an effect to set `blockEditingMode` for its inner blocks to the correct thing. 

This PR moves that calculation to a selector so that it plays better with other modes.

## Why?
The `getBlockEditingMode` has some internal logic for different editing modes (like `zoom-out` and the content only behavior in #65204) that can override the actual block editing state for a pattern's inner blocks.

## How?
Moving the state to the selector it should be possible to make the modes work together with pattern overrides in the expected way.

## Testing Instructions
There's one use case that I can see that doesn't seem to work, but there might be others
1. Add a synced pattern with overrides into the content block of a page in the site editor
2. Wrap that synced pattern in a group
3. Switch to 'Edit' mode
4. Try editing the overrides - notice they're locked.

#### Before
An pattern that's a child of a 'section' group has overrides that can't be edited in Edit Mode (even though they're 'content')
Screenshot 2024-09-17 at 4 33 22 pm

#### After
The pattern overrides can be selected and edited in Edit Mode
Screenshot 2024-09-17 at 4 32 24 pm

### CODE REVIEW

File: packages/block-editor/src/store/selectors.js
Comment (talldan) [inline packages/block-editor/src/store/selectors.js]: I haven't tested this with 'zoom-out' mode yet, so it might need some adjustment.

File: packages/block-editor/src/store/private-selectors.js
Comment (talldan) [inline packages/block-editor/src/store/private-selectors.js:696]: Might be an idea to make this one memoized, and also check that the memoization of selectors that use `getBlockEditingMode` is still correct 🤔

File: packages/block-editor/src/store/test/selectors.js
Comment (talldan) [inline packages/block-editor/src/store/test/selectors.js:3082]: This test was failing because it tries to insert a block within a child of a `core/block` block. This wouldn't be possible by a user in trunk (as it's disallowed by the block's 'edit') and it also failed in unit tests in this PR now that the block editing mode is enforced at selector level.

I've switched it to a group instead.

File: packages/block-editor/src/store/selectors.js
Comment (getdave) [reply packages/block-editor/src/store/selectors.js]: https://github.com/user-attachments/assets/0e8456ad-7271-43f2-b542-6bd0a6bc5b36

I tested it. The only quirk is that the overridable fields highlight when the block is clicked.

**Comment by draganescu** [comment]
This test very well for me. I also noticed what @getdave mentioned that when zoomed out to compose patterns the overrides highlight on click - it'd be great to remove that then but this is out of the scope of this - this fixes editing overrides in the new edit mode.

Great work, will you merge this @talldan ?

File: packages/block-editor/src/store/selectors.js
Comment (getdave) [reply packages/block-editor/src/store/selectors.js]: The code for flashing blocks is part of the props passed to `useFlashEditableBlocks` hook in `packages/block-editor/src/components/block-list/use-block-props/index.js`

Comment (talldan) [reply packages/block-editor/src/store/selectors.js]: This one is a feature to show which parts of the pattern are editable, I think @noisysocks worked on it.

Comment (getdave) [reply packages/block-editor/src/store/selectors.js]: It would actually be a good feature _if_ we could edit those fields in zoom out view. In the future we _might_ want to allow that.

Comment (talldan) [reply packages/block-editor/src/store/selectors.js]: Ah, I see, I didn't realize they weren't editable. I've pushed a commit that adjusts that.

**Comment by talldan** [comment]
> This test very well for me. I also noticed what @getdave mentioned that when zoomed out to compose patterns the overrides highlight on click - it'd be great to remove that then but this is out of the scope of this - this fixes editing overrides in the new edit mode.

That should be fixed by the latest commit.

@youknowriad Are you happy for this to merged into your branch? I can also try basing it on trunk, though the things it fixes are more relevant to that branch.

**Comment by youknowriad** [comment]
@talldan Let's keep it separate. I believe my PR is an approval away from being merged. I think it's better for the history to keep these two separate. We can merge this right after.

**Comment by talldan** [comment]
I've rebased this PR. I'm not sure how relevant it still is, I've fallen a bit behind all the work that's been happening lately.

File: packages/block-editor/src/store/selectors.js
Comment (getdave) [reply packages/block-editor/src/store/selectors.js]: This is our first "fork" situation. When https://github.com/WordPress/gutenberg/pull/65702 is merged we will want Gutenberg `trunk` to flash blocks but WP 6.7 _not_ to flash blocks (because in 6.7 you can't edit sub content) 😬

Comment (talldan) [reply packages/block-editor/src/store/selectors.js]: 🤔 Maybe we can utilize `IS_GUTENBERG_PLUGIN` for the bit that's different.

File: packages/block-editor/src/store/private-selectors.js
Comment (youknowriad) [inline packages/block-editor/src/store/private-selectors.js]: Why is this a separate selector not part of the more generic `getBlockEditingMode`?

Comment (talldan) [reply packages/block-editor/src/store/private-selectors.js]: I thought it might be more readable. I don't have a strong preference though, I've pushed a commit that makes it one selector.

Comment (youknowriad) [reply packages/block-editor/src/store/private-selectors.js]: Ah, making it separate for clarity like an internal function is fine but the problem with it exposed as a separate selector is the fact that the "framework" when it needs to know the block's editing mode has to switch between two different selectors.

File: packages/block-library/src/block/edit.js
Comment (youknowriad) [inline packages/block-library/src/block/edit.js:215]: Nice to see this gone :)

**Review by draganescu — APPROVED** [review]
Ha! This PR now fixes a different problem: initially the problem was that editable parts of synced patterns were not editable in write mode. However that is fixed now in `trunk`.

But there is another bug in `trunk` that this fixes: in `trunk` in write mode you can edit uneditable parts of a synced pattern. With this applied that is fixed.

**Comment by tyxla** [comment]
Is it possible this PR introduced a regression to typing metrics:

* https://www.codevitals.run/project/gutenberg/type
* https://www.codevitals.run/project/gutenberg/site-editor-type

Screenshot 2024-10-08 at 14 38 48

Screenshot 2024-10-08 at 14 38 59

**Comment by talldan** [comment]
> Is it possible this PR introduced a regression to typing metrics:

@tyxla Thanks for catching. Yes, from what I can tell `getBlockEditingMode` is called quite a lot, and I should've checked the changes before merging. I'll try pushing up a PR to improve things, but I also think it might be good to revert for that kind of spike!

**Comment by talldan** [comment]
Reverted in #65953

**Comment by getdave** [comment]
I wonder if we could/should look at why `getBlockEditingMode` is called so frequently? Is it happening on block "hover" or anything like that?

**Comment by youknowriad** [comment]
It seems understandable that `getBlockEditingMode` gets called often because it's what controls how each block is interacted with. That said, if the selector becomes too slow, we should consider denormalizing it like we do for block tree.

In other words, add a "computed" `blockEditingMode` state to the reducer that is updated on each action and the selector would just return that value. It's a duplicated value but it's an approach that is reasonable when the performance gains are obvious.

**Pull request merged**
